packages/api/lib/ingest.js

```diff
@@ -0,0 +1,157 @@
+'use strict';
+
+const path = require('path');
+
+const Lambda = require('@cumulus/aws-client/Lambda');
+const StepFunctions = require('@cumulus/aws-client/StepFunctions');
+const { getKnexClient, GranulePgModel } = require('@cumulus/db');
+
+const { deconstructCollectionId } = require('./utils');
+const { Granule, Rule } = require('../models');
+
+/**
+   * Set the Dynamo and PG Granule "status" field to "running"
+   *
+   * @private
+   * @param {Knex} knex - DB client
+   * @param {string} granuleId - the granule's ID
+   * @param {GranulePgModel} granulePgModel - Postgres Granule model
+   * @param {Granule} granuleModel - API Granule model
+   * @returns {Promise<undefined>} - undefined
+   */
+async function _updateGranuleStatus(
+  knex,
+  granuleId,
+  granulePgModel,
+  granuleModel
+) {
+  await granuleModel.updateStatus({ granuleId: granuleId }, 'running');
+  const pgGranuleToUpdate = await granulePgModel.getByGranuleId(
+    knex,
+    granuleId
+  );
+  await granulePgModel.upsert(
+    knex,
+    {
+      ...pgGranuleToUpdate,
+      status: 'running',
+    }
+  );
+}
+
+/**
+   * start the re-ingest of a given granule object
+   *
+   * @param {Object} params
+   * @param {Object} params.reingestParams - the granule object with additional params
+   * @param {string} [params.asyncOperationId] - specify asyncOperationId origin
+   * @param {Granule} [params.granuleModel] - API Granule model (optional, for testing)
+   * @param {GranulePgModel} [params.granulePgModel] - Postgres Granule model
+   * (optional, for testing)
+   * @returns {Promise<undefined>} - undefined
+   */
+async function reingestGranule({
+  reingestParams,
+  asyncOperationId = undefined,
+  granuleModel = new Granule(),
+  granulePgModel = new GranulePgModel(),
+}) {
+  const knex = await getKnexClient();
+  const executionArn = path.basename(reingestParams.execution);
+
+  const executionDescription = await StepFunctions.describeExecution({ executionArn });
+  const originalMessage = JSON.parse(executionDescription.input);
+
+  const { name, version } = deconstructCollectionId(reingestParams.collectionId);
+
+  const lambdaPayload = await Rule.buildPayload({
+    workflow: originalMessage.meta.workflow_name,
+    meta: originalMessage.meta,
+    cumulus_meta: {
+      cumulus_context: {
+        reingestGranule: true,
+        forceDuplicateOverwrite: true,
+      },
+    },
+    payload: originalMessage.payload,
+    provider: reingestParams.provider,
+    collection: {
+      name,
+      version,
+    },
+    queueUrl: reingestParams.queueUrl,
+    asyncOperationId,
+  });
+
+  // FUTURE This would ideally not be necessary
+  await _updateGranuleStatus(
```

Comment:
We should ensure this is unit tested, either at the lib level *or* the endpoint level.

Reply:
Endpoint test has this, consider moving/duping that test here.

Reply:
Yeah, but.... 'stubs stink'. 

Try https://github.com/nasa/cumulus/pull/2463 on for size.